src/utils/AuthProvider.tsx

```diff
@@ -0,0 +1,116 @@
+export interface AuthContextType {
+  session: Session | undefined;
+  userId: UUID | undefined;
+  userEmail: string | undefined;
+  signIn: (email: string, password: string) => Promise<AuthResponse>;
+  signOut: () => Promise<AuthError | null>;
+  signUp: (
+    email: string,
+    password: string,
+    options: object,
+  ) => Promise<AuthResponse>;
+}
+
+export const AuthContext = createContext<AuthContextType | undefined>(
+  undefined,
+);
+
+// TODO: we might need error handling here?
+export const useAuth = () => useContext(AuthContext);
+
+export default function AuthProvider({
+  children,
+}: {
+  children: React.ReactNode;
+}) {
+  const [session, setSession] = useState<Session | undefined>(undefined);
+  const [userId, setUserId] = useState<UUID | undefined>(undefined);
+  const [userEmail, setUserEmail] = useState<string | undefined>(undefined);
+
+  // set the session, userId, and userEmail
+  const setAll = (newSession: Session | undefined) => {
+    if (!newSession) return;
+    setSession(newSession);
+    if (newSession?.user?.id) {
+      const sessionUserId = newSession.user.id as UUID;
+      const sessionUserEmail = newSession.user.email;
+      setUserId(sessionUserId);
+      setUserEmail(sessionUserEmail);
+    }
+  };
+
+  // on page load, check if there's a session and set it
+  useEffect(() => {
+    supabase.auth.getSession().then(({ data: { session: newSession } }) => {
+      setAll(newSession != null ? newSession : undefined);
+    });
+
+    supabase.auth.onAuthStateChange((_event, newSession) => {
+      setAll(newSession != null ? newSession : undefined);
+    });
+  }, []);
+
+  // sign in and set the session, userId, and userEmail
+  const signIn = async (email: string, password: string) => {
+    const value = await supabase.auth.signInWithPassword({
+      email,
+      password,
+    }); // will trigger onAuthStateChange to update the session
+
+    setAll(value.data.session ? value.data.session : undefined);
+
+    return value;
+  };
+
+  // sign out and clear the session, userId, and userEmail
+  const signOut = async () => {
+    const { error } = await supabase.auth.signOut();
+
+    setSession(undefined);
+    setUserId(undefined);
+    setUserEmail(undefined);
+
+    return error;
+  };
+
+  // sign up and set the session, userId, and userEmail
+  const signUp = async (email: string, password: string, options: object) => {
+    console.log('signUp', email, password, options);
+    const value = await supabase.auth.signUp({
+      email,
+      password,
+      options: {
+        ...options,
+      },
+    }); // will trigger onAuthStateChange to update the session
+
+    setAll(value.data.session ? value.data.session : undefined);
+    return value;
+  };
+
+  const authContextValue = useMemo(
+    () => ({
+      session,
+      userId,
+      userEmail,
+      signIn,
+      signOut,
+      signUp,
+    }),
+    [session, userId, userEmail, signIn, signOut, signUp],
+  );
```

Comment:
thanks jinkang!! i think i'm a little confused about the syntax of the fix lol so let's discuss on sunday

Reply:
In hindsight, I doubt we'd need to use a `useCallback` hook at all, since they don't return a value for the memoization to be helpful. We could probably just define the functions as we would for normal handlers and toss that along the other values (which may or may not be wrapped in `useMemo`).

Reply:
(Recap of what we discussed this during MPH) It seems like using this causes a TS compiler error, so I'm just using what we already had for right now. but @jinkang-0 @varortz curious if you have other approaches you'd like to suggest?

Reply:
Update: I wrapped the auth functions in `useCallback`s and left everything else as is, and this seems to get rid of all TS compiler errors -- but let me know what y'all think!!